### PR TITLE
[Feature] support mmyolo deployment

### PR DESCRIPTION
**Pull request by hanrui1sensetime**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Support mmyolo deploy.

## Modification

Add deploy folder in mmyolo/mmyolo.

## BC-breaking (Optional)

None.

## Use cases (Optional)

None.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDetection or MMClassification.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmyolo/deploy/codebase/mmyolo/deploy/__init__.py [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from .object_detection import MMYolo, YoloObjectDetection
```
Comment (PeterH0323) [inline mmyolo/deploy/codebase/mmyolo/deploy/__init__.py:2]:
```suggestion
from .object_detection import MMYOLO, YOLOObjectDetection
```

File: mmyolo/deploy/codebase/mmyolo/deploy/__init__.py [hunk 2]
```diff
@@ -0,0 +1,4 @@
...
+
+__all__ = ['MMYolo', 'YoloObjectDetection']
```
Comment (PeterH0323) [inline mmyolo/deploy/codebase/mmyolo/deploy/__init__.py:4]:
```suggestion
__all__ = ['MMYOLO', 'YOLOObjectDetection']
```

File: mmyolo/deploy/codebase/mmyolo/deploy/object_detection.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+from mmdeploy.codebase.base import CODEBASE, MMCodebase
+from mmdeploy.codebase.mmdet.deploy import ObjectDetection
+from mmdeploy.utils import Codebase, Task
+
+MMYOLO_TASK = Registry('mmyolo_tasks')
+
+
+@CODEBASE.register_module(Codebase.MMYOLO.value)
+class MMYolo(MMCodebase):
```
Comment (PeterH0323) [inline mmyolo/deploy/codebase/mmyolo/deploy/object_detection.py:15]:
```suggestion
class MMYOLO(MMCodebase):
```

File: mmyolo/deploy/codebase/mmyolo/deploy/object_detection.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+            if meta is not None:
+                return meta
+        if hasattr(dataset_cls, 'METAINFO'):
+            return dataset_cls.METAINFO
+
+    return None
+
+
+@MMYOLO_TASK.register_module(Task.OBJECT_DETECTION.value)
+class YoloObjectDetection(ObjectDetection):
```
Comment (PeterH0323) [inline mmyolo/deploy/codebase/mmyolo/deploy/object_detection.py:57]:
```suggestion
class YOLOObjectDetection(ObjectDetection):
```

File: mmyolo/deploy/codebase/mmyolo/models/dense_heads/yolov5_head.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+from mmdeploy.codebase.mmdet import get_post_processing_params
+from mmdeploy.codebase.mmdet.models.layers import multiclass_nms
+from mmdeploy.core import FUNCTION_REWRITER
+
+
+@FUNCTION_REWRITER.register_rewriter(
+    func_name='mmyolo.models.dense_heads.yolov5_head.'
+    'YOLOv5Head.predict_by_feat')
+def yolov5_head__predict_by_feat(ctx,
```
Comment (PeterH0323) [inline mmyolo/deploy/codebase/mmyolo/models/dense_heads/yolov5_head.py:18]:
```suggestion
def yolov5_head_predict_by_feat(ctx,
```

**Review by PeterH0323 — CHANGES_REQUESTED** [review]

**Comment by PeterH0323** [comment]
Hi @hanrui1sensetime 

Thx for your contribute !

Pls merge to `dev` branch. 😄

File: mmyolo/deploy/codebase/mmyolo/models/dense_heads/yolov5_head.py (hunk 5 above)
Comment (hanrui1sensetime) [reply mmyolo/deploy/codebase/mmyolo/models/dense_heads/yolov5_head.py:18]: > 

In mmdeploy, we define the rewriter function name as: `model__function__backend`, so I think `yolov5_head__predict_by_feat` is okay.

Comment (cydiachen) [inline mmyolo/deploy/codebase/mmyolo/models/dense_heads/yolov5_head.py:18]: Seems as a typo

Comment (hanrui1sensetime) [reply mmyolo/deploy/codebase/mmyolo/models/dense_heads/yolov5_head.py:18]: > Seems as a typo

No, in mmdeploy, we define the rewriter function name as: model__function__backend, so I think yolov5_head__predict_by_feat is okay.

File: configs/deploy/mmyolo/detection_onnxruntime_dynamic.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
+onnx_config = dict(
```
Comment (RunningLeon) [inline configs/deploy/mmyolo/detection_onnxruntime_dynamic.py:1]: may change all configs to `configs/deploy/`

File: mmyolo/deploy/__init__.py [hunk 7]
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from mmdeploy.codebase.base import BaseTask, MMCodebase, get_codebase_class
+
+from .models import *  # noqa: F401,F403
+from .object_detection import MMYOLO, YOLOObjectDetection
+
+__all__ = [
+    'MMCodebase', 'BaseTask', 'get_codebase_class', 'MMYOLO',
```
Comment (RunningLeon) [inline mmyolo/deploy/__init__.py:8]: Could remove 'BaseTask', 'get_codebase_class'

**Comment by RunningLeon** [comment]
@hanrui1sensetime could fix lint

File: configs/deploy/detection_onnxruntime_dynamic.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
+onnx_config = dict(
```
Comment (RunningLeon) [inline configs/deploy/detection_onnxruntime_dynamic.py:1]: same codebase config and onnx config appear in multiple configs, could simply by making  base static and dynamic configs.

**Comment by RunningLeon** [comment]
@hanrui1sensetime pls fix lint error

**Comment by RunningLeon** [comment]
@hhaAndroid Any comment on this PR?

**Comment by hhaAndroid** [comment]
@RunningLeon Please sync the latest dev, ci failed

File: .pre-commit-config.yaml [hunk 9]
```diff
@@ -1,7 +1,7 @@
 exclude: ^tests/data/
 repos:
   - repo: https://github.com/PyCQA/flake8
-    rev: 3.8.3
+    rev: 5.0.4
```
Comment (hhaAndroid) [inline .pre-commit-config.yaml:4]: has been fixed

File: tests/test_deploy/data/model.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+# model settings
+data_preprocessor = dict(
```
Comment (hhaAndroid) [inline tests/test_deploy/data/model.py:3]: Should it be changed to yolov5 configuration？

File: .pre-commit-config.yaml (hunk 9 above)
Comment (hanrui1sensetime) [reply .pre-commit-config.yaml:4]: lint will failed if we use flake8 == 3.8.3. It conflicts with `importlib_metadata`
https://github.com/python/importlib_metadata/issues/406

**Comment by RunningLeon** [comment]
@hanrui1sensetime Could you fix ci?

**Review by RunningLeon — APPROVED** [review]
LGTM

**Review by hhaAndroid — APPROVED** [review]

**Pull request merged**